### DISCUSSION TITLE
ST7701S Support

### DISCUSSION THREAD

**Discussion by flykarlos** [Bodmer/TFT_eSPI#1055] [New processor/display support or capability request] [answered]
Dear Bodmer

I am writing here with request to add support for displays with ST7701S driver. I am approaching to realization of my project which will be using display equipped with such driver IC. I ordered this one (SPI version). It is 480x480 resolution.

http://surenoo.tech/download/04_STP/R0210/STP480480_R021A2.pdf?spm=a2g0o.detail.1000023.17.454942c2mfqUMx&file=STP480480_R021A2.pdf

I am amazed, how great is Your TFT_eSPI library. Also, I am aware of how much work it goes to maintain and provide such support, so I would like to ask you if you have any donation profile, pay pal etc? Your work has to be appreciated more than saying thanks every time You help.
Also, please send me email with delivery address so I can order same display for You for test purposes.
 Thank You very much!
Karol

**Comment by Bodmer** [comment, marked as answer]
Hi, I have a lot of projects on the go at the moment so unfortunately I do not have the time to help you.

The library does allow new displays to be added, but this is not something I have time to do. Thanks for your offer and kind comments about the library.

**Comment by flykarlos** [comment]
Thanks for answer. Do You have any guide how to create a driver? I went through existing drivers and looks like they all follow some "schematic". Do You know any document to help me understand how should I do it? or I just have to study driver IC manual and successively create new driver based on drivers already existing?  Maybe there's no single and simple answer.

**Comment by Bodmer** [comment]
To add a new display 3 new files need to be created[ here](https://github.com/Bodmer/TFT_eSPI/tree/master/TFT_Drivers).  xxxx_Init.h, xxxx_Defines.h and xxxx_Rotation.h

The init sequence is critical and needs to set the display to 4 wire and 16 bit colour. The Init sequences can often be found on the web using a Google search, or may be provided by the display supplier.

Then a reference needs to be allocated [here](https://github.com/Bodmer/TFT_eSPI/blob/master/User_Setup_Select.h#L188).

And links created, [here](https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_eSPI.cpp#L599) and [here](https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_eSPI.cpp#L599).

**Reply by flykarlos** [reply]
Thanks a lot, I will try to do that as soon as I get my display from China. 
In case of succes i will not hesitate to share driver files. Before that I will most probably ask some more questions :)
Thanks!

**Reply by flykarlos** [reply]
Hi Bodmer

Do You have any advices about creating driver for three wire SPI? my display is missing DC pin and Datasheet mentions 3SPI..
![Zrzut ekranu 2021-03-18 o 20 49 37](https://user-images.githubusercontent.com/18057422/111688409-7900d100-882b-11eb-8e1b-308b7e1fa313.png)

Good info is that I received init code from manufacturer:

```
HW_Reset();   			//VBP>=11, VFP>=7

Delay(120);

SPI_WriteComm(0xFF);
SPI_WriteData(0x77);
SPI_WriteData(0x01);
SPI_WriteData(0x00);
SPI_WriteData(0x00);
SPI_WriteData(0x10);

SPI_WriteComm(0xC0);
SPI_WriteData(0x3B);//Scan line	
SPI_WriteData(0x00);

SPI_WriteComm(0xC1);
SPI_WriteData(0x0B);	//VBP
SPI_WriteData(0x02);

SPI_WriteComm(0xC2);
SPI_WriteData(0x07);
SPI_WriteData(0x02);

SPI_WriteComm(0xCC);
SPI_WriteData(0x10);

//SPI_WriteComm(0xCD);//RGB format
//SPI_WriteData(0x08);

SPI_WriteComm(0xB0); // IPS   
SPI_WriteData(0x00); // 255 
SPI_WriteData(0x11); // 251    
SPI_WriteData(0x16); // 247  down
SPI_WriteData(0x0e); // 239    
SPI_WriteData(0x11); // 231    
SPI_WriteData(0x06); // 203    
SPI_WriteData(0x05); // 175 
SPI_WriteData(0x09); // 147    
SPI_WriteData(0x08); // 108    
SPI_WriteData(0x21); // 80  
SPI_WriteData(0x06); // 52   
SPI_WriteData(0x13); // 24    
SPI_WriteData(0x10); // 16    
SPI_WriteData(0x29); // 8    down
SPI_WriteData(0x31); // 4    
SPI_WriteData(0x18); // 0   

SPI_WriteComm(0xB1);//  IPS	   
SPI_WriteData(0x00);//  255 
SPI_WriteData(0x11);//  251
SPI_WriteData(0x16);//  247   down
SPI_WriteData(0x0e);//  239
SPI_WriteData(0x11);//  231
SPI_WriteData(0x07);//  203    
SPI_WriteData(0x05);//  175
SPI_WriteData(0x09);//  147  
SPI_WriteData(0x09);//  108  
SPI_WriteData(0x21);//  80 
SPI_WriteData(0x05);//  52   
SPI_WriteData(0x13);//  24 
SPI_WriteData(0x11);//  16 
SPI_WriteData(0x2a);//  8  down 
SPI_WriteData(0x31);//  4  
SPI_WriteData(0x18);//  0  

SPI_WriteComm(0xFF);
SPI_WriteData(0x77);
SPI_WriteData(0x01);
SPI_WriteData(0x00);
SPI_WriteData(0x00);
SPI_WriteData(0x11);

SPI_WriteComm(0xB0);  //VOP  3.5375+ *x 0.0125
SPI_WriteData(0x6d);  //5D
 
SPI_WriteComm(0xB1); 	//VCOM amplitude setting  
SPI_WriteData(0x37);  //
  
SPI_WriteComm(0xB2); 	//VGH Voltage setting  
SPI_WriteData(0x81);	//12V

SPI_WriteComm(0xB3);
SPI_WriteData(0x80);

SPI_WriteComm(0xB5); 	//VGL Voltage setting  
SPI_WriteData(0x43);	//-8.3V

SPI_WriteComm(0xB7);
SPI_WriteData(0x85);

SPI_WriteComm(0xB8);
SPI_WriteData(0x20);

SPI_WriteComm(0xC1);
SPI_WriteData(0x78);

SPI_WriteComm(0xC2);
SPI_WriteData(0x78);

SPI_WriteComm(0xD0);
SPI_WriteData(0x88);

SPI_WriteComm(0xE0);
SPI_WriteData(0x00);
SPI_WriteData(0x00);
SPI_WriteData(0x02);

SPI_WriteComm(0xE1);
SPI_WriteData(0x03);	
SPI_WriteData(0xA0);
SPI_WriteData(0x00);	
SPI_WriteData(0x00);
SPI_WriteData(0x04);	
SPI_WriteData(0xA0);
SPI_WriteData(0x00);	
SPI_WriteData(0x00);	
SPI_WriteData(0x00);
SPI_WriteData(0x20);
SPI_WriteData(0x20);

SPI_WriteComm(0xE2);
SPI_WriteData(0x00);	
SPI_WriteData(0x00);	  
SPI_WriteData(0x00);	
SPI_WriteData(0x00);	
SPI_WriteData(0x00);	
SPI_WriteData(0x00);	
SPI_WriteData(0x00);	
SPI_WriteData(0x00);	
SPI_WriteData(0x00);	
SPI_WriteData(0x00);	
SPI_WriteData(0x00);	
SPI_WriteData(0x00);	
SPI_WriteData(0x00);

SPI_WriteComm(0xE3);
SPI_WriteData(0x00);
SPI_WriteData(0x00);
SPI_WriteData(0x11);
SPI_WriteData(0x00);

SPI_WriteComm(0xE4);
SPI_WriteData(0x22);
SPI_WriteData(0x00);

SPI_WriteComm(0xE5);		
SPI_WriteData(0x05);	
SPI_WriteData(0xEC);	
SPI_WriteData(0xA0);
SPI_WriteData(0xA0);
SPI_WriteData(0x07);	
SPI_WriteData(0xEE);	
SPI_WriteData(0xA0);
SPI_WriteData(0xA0);
SPI_WriteData(0x00);
SPI_WriteData(0x00);	
SPI_WriteData(0x00);
SPI_WriteData(0x00);
SPI_WriteData(0x00);
SPI_WriteData(0x00);	
SPI_WriteData(0x00);
SPI_WriteData(0x00);

SPI_WriteComm(0xE6);
SPI_WriteData(0x00);
SPI_WriteData(0x00);
SPI_WriteData(0x11);
SPI_WriteData(0x00);

SPI_WriteComm(0xE7);
SPI_WriteData(0x22);
SPI_WriteData(0x00);

SPI_WriteComm(0xE8);		
SPI_WriteData(0x06);	
SPI_WriteData(0xED);	
SPI_WriteData(0xA0);
SPI_WriteData(0xA0);
SPI_WriteData(0x08);	
SPI_WriteData(0xEF);	
SPI_WriteData(0xA0); 
SPI_WriteData(0xA0);
SPI_WriteData(0x00);	
SPI_WriteData(0x00);	
SPI_WriteData(0x00);
SPI_WriteData(0x00);
SPI_WriteData(0x00);	
SPI_WriteData(0x00);	
SPI_WriteData(0x00);
SPI_WriteData(0x00);

SPI_WriteComm(0xEB);
SPI_WriteData(0x00); 	
SPI_WriteData(0x00);
SPI_WriteData(0x40);
SPI_WriteData(0x40);
SPI_WriteData(0x00);
SPI_WriteData(0x00);
SPI_WriteData(0x00);  

SPI_WriteComm(0xED);  
SPI_WriteData(0xFF); 
SPI_WriteData(0xFF);  
SPI_WriteData(0xFF); 	
SPI_WriteData(0xBA); 		
SPI_WriteData(0x0A); 	
SPI_WriteData(0xBF); 	
SPI_WriteData(0x45); 	
SPI_WriteData(0xFF); 
SPI_WriteData(0xFF);  
SPI_WriteData(0x54); 	
SPI_WriteData(0xFB); 	
SPI_WriteData(0xA0); 	
SPI_WriteData(0xAB); 	
SPI_WriteData(0xFF); 
SPI_WriteData(0xFF); 
SPI_WriteData(0xFF); 

SPI_WriteComm(0xEF);
SPI_WriteData(0x10); 
SPI_WriteData(0x0D); 
SPI_WriteData(0x04); 
SPI_WriteData(0x08); 
SPI_WriteData(0x3F); 
SPI_WriteData(0x1F);

SPI_WriteComm(0xFF);
SPI_WriteData(0x77);
SPI_WriteData(0x01);
SPI_WriteData(0x00);
SPI_WriteData(0x00);
SPI_WriteData(0x13);

SPI_WriteComm(0xEF);
SPI_WriteData(0x08);

SPI_WriteComm(0xFF);
SPI_WriteData(0x77);
SPI_WriteData(0x01);
SPI_WriteData(0x00);
SPI_WriteData(0x00);
SPI_WriteData(0x00);

SPI_WriteComm(0x11);

Delay(120);

SPI_WriteComm(0x29);

SPI_WriteComm(0x36);
SPI_WriteData(0x08);

SPI_WriteComm(0x3A);
SPI_WriteData(0x70);

```

thanks for suggestions!

**Reply by Bodmer** [reply]
The complication with the 3 wire displays is that the Data/Command (DC) bit must be sent as a 9th bit for every byte sent. This is possible but is an extra complicationsince the data stream has to be divided up and shifted to make use of the 8/32 bit SPI output registers/buffers. I have avoided since there are plenty of supported 4 wire displays available.

There are Arduino graphics libraries that do support 3 wire displays, but it is not a feature I will be adding.

**Reply by flykarlos** [reply]
Thank You for answer

Just after posting my question I realized that You actually mentioned on main page that 3 wire SPI is not supported. 
I agree that many cheap displays are available, but it is not an option in my case, because I am limited to round shape displays  due to specification of my project. There are a number of round displays available, but if somebody is looking on something bigger than 2 inch in diameter with SPI support currently there is only ST7701S based display.
Whats funny is that driver IC itself supports 4 wire SPI, but DC pin is not routed on ribbon cable..
Anyway, thanks for Your hard work!

**Reply by flykarlos** [reply]
Sorry for bothering you again. I am in process of learning and understanding display interfaces. One thing bothers me..
I fully understand that Your library does not support 3 wire SPI. Thats perfectly ok. But I got "addicted" to this library which suits almost all my needs, so I would stick to it with future projects. So I am looking for a way to use 3 wire SPI.
During surfing internet for more info about usage of SPI+RGB interface I found this:

_`The 3-wire Serial Peripheral Interface (SPI) and the 16/18/24-bit RGB parallel interface are used in a sequence to communicate data to the display. The SPI interface is first used to initialize the display parameters through command registers. Once initialized, the display can receive data over the RGB parallel interface at a higher speed. This data can be sent in segments of 16, 18 or 24-bit lengths depending on what is specified by the SPI initialization commands.`_

Sorry for pretty noob question, but.. theoretically.. would it be possible to initialize display using 3 wire SPI using some other library or just raw code, and then drive same display using 18 bit RGB making use of Your library? Is Your library limited to 8 Bit parallel?

**Reply by Bodmer** [reply]
The RGB interface is intended to be driven by a dedicated hardware graphics chip and requires other control signals (e.g. dot clock at 6.35MHz, v sync and h sync etc) these have very critical timing requirements that would be tricky to generate with a microcontroller.

**Reply by flykarlos** [reply]
thanks for reply, I thought so that it is not that simple. I have to stick with other display models. Thanks again :)

**Comment by mm108** [comment]
@flykarlos hi, any (further) luck with getting ST7701S to work? I am looking at a 2.8 inch TFT LCD ,480x640 with ST7701S driver IC. I am planning to get a bunch of random boards ( but only those with higher resolution ).

The other one I am getting is driven by HX8357 (3.2 inch/480x320). This I think some folks have been able to get it to work so I haven't researched that much yet. This one I think should not be a problem.

And thanks to you all wonderful folks who have built this library and contributed to it's advancements.

**Reply by flykarlos** [reply]
> @flykarlos hi, any (further) luck with getting ST7701S to work? I am looking at a 2.8 inch TFT LCD ,480x640 with ST7701S driver IC. I am planning to get a bunch of random boards ( but only those with higher resolution ).
> 
> 
> 
> The other one I am getting is driven by HX8357 (3.2 inch/480x320). This I think some folks have been able to get it to work so I haven't researched that much yet. This one I think should not be a problem.
> 
> 
> 
> And thanks to you all wonderful folks who have built this library and contributed to it's advancements.

Hi m-menon
Unfortunately I was not able to modify Bodmer's library or find any other library currently supporting 9bit SPI ST7701s display. 
There is a library called Arduino_GFX  whose owner says about adding optional support to 9bit SPI in the future. 
 I chosen another approach and I'm using raspberry pi together with hyperpixel 2.1 round display. I wanted to stick with ESP32 /ESP32S and 480x480 round display, but I had to switch to RPI to be able to continue my project with round display.

My opinion about Bodmer's library stays unchanged- Greatest library that I ever used, and best support I have ever received.

Thanks Bodmer once again!

**Reply by mm108** [reply]
Thanks @flykarlos for the update. I checked the [Arduino_GFX](https://github.com/moononournation/Arduino_GFX) and it's work in progress but it does support 9 bit SPI and a bunch of other chips. One of the first things with [Arduino_GFX](https://github.com/moononournation/Arduino_GFX) is the (slower) speed. It does support a bunch of driver ICs and it does include couple of those that I am interested in. 

Also another "temporary" option for those really in need of high res displays on their ESP32, there is this option on [Tindie](https://www.tindie.com/products/prodbld/40-inch-ft81x-tft-lcd-display-module-for-arduino/). It is out of stock though and kinda expensive for those who just want a high res display on esp32 without the other bells and whistles.

As for Bodmer's TFT_eSPI, I have the same opinion as you. It is one of THE BEST libraries out there and the support from Bodmer has been nothing short of amazing. TFT_eSPI is my go-to TFT display library unless there is a limitation where a solution, workaround or hack is impossible. I can't thank Bodmer enough for TFT_eSPI.

And ya thanks once again for the update.

**Comment by projectitis** [comment]
I've got a device with the ST7701S display working with a sample project using ESP-IDF ([this one](https://github.com/wireless-tag-com/ZX2D10GE01R-V4848), in fact).
I'm now converting to PlatformIO and Arduino using TFT_eSPI, and am adding support for ST7701S to TFT_eSPI.
I have the startup sequence, the rotation and _most_ of the commands/defines... 

However, ST7701S does not have GRAM, and as such does not have a CASET, PASET, RAMWR or RAMRD command. These are required in `ST7701S_Defines.h`.
@Bodmer, is there another display you support that doesn't have GRAM that I can copy as an example?
If this is the first GRAM-less display, I suspect there will be lots of other changes required to support it, and that would be beyond me.

**Reply by DrJaymz** [reply]
I am using the lilygo-t-rgb with this display.  I am also using the TFT_eSPI library. What I am doing is preparing a main sprite and then pushing the memory to the panel object.

    // finally output the image
    panel.pushColors(0, 0, WIDTH, HEIGHT, (uint16_t *)fbuff.getPointer());

I don't really know how efficient doing it this way is, but I also didn't want to learn yet another framework.  Copying memory feels like it should be reasonably quick and I find it takes me about 70ms to do that.  So its a bit like an unnecessary double buffer.  This means that I can use the functions in the library and presumably there will be some limitations but I achived what I set out to.

If there's a better solution - I would love to hear it.

**Reply by projectitis** [reply]
@DrJaymz are you sure you are using TFT_eSPI? If so, which TFT_Driver are you using - I can't find a matching driver for the ST7701s, or a user-setup, and same for the lilygo-t-rgb.

I have my code working with a different display library (TFT library for Arduino) which supports the ST7701s but I haven't been able to get it to work with TFT_eSPI.

Sorry - reading your post again it sounds like you are using a combination of libraries, correct? TFT_eSPI just for its drawing routines, and then pushing the buffer to the TFT using another library?

**Reply by Sturmbock** [reply]
I have a related problem with a ST7701 display, but have a solution to it in part. I am using an Adafruit Qualia ESP32S3 card to interface to it, so I can use the TFT_eSPI library to manipulate sprites and push them to memory (pushSprite). I then use GFX (draw16bitBeRGBBitmap) to draw them to the display. The bit I can’t find a solution to is, I normally use the TFT library (pushRotated) to push the needle on a gauge at a specific angle to the display. The GFX library does not seem to support this, any ideas most welcome.

**Reply by bajacondor** [reply]
@DrJaymz , Can you provide a bit more context/example of your approach?  I'm trying to do something similar.  I have a bus to the panel working with esp_display_panel, I would like to write a TFT_eSPI buffer directly to that panel so that I can draw with TFT_eSPI and write to esp_display_panel's bus.  Is that what you are doing?

**Reply by DrJaymz** [reply]
@bajacondor  Sure.  Basically you have an instance of TFT_eSPI just to create a sprite which you draw on as you normally would.  Then you copy that sprite to the panel its a memory copy so its quite fast.

```
LilyGo_RGBPanel panel;

// Initialize T-RGB, 
panel.begin(LILYGO_T_RGB_ORDER_BGR));
   
//create an instance of TFT_eSPI
TFT_eSPI tft = TFT_eSPI();

//create a sprite
TFT_eSprite fbuff= TFT_eSprite(&tft);

// do all your stuff with the sprite (keep it global so the memory stays static
fbuff.createSprite(WIDTH, HEIGHT);
fbuff.setColorDepth(16);

//draw whatever...
fbuff.<draw whatever you like using TFT_eSPI>
sprintf(strBuff, "%0.0f", vs);
fbuff.drawString(strBuff, vsiLeft + 8, vsiBug, GFXFF);
fbuff.drawWedgeLine(centerX, centerY, posX, posY, 8, 4, needle_colour);

// finally output the image - this copies the sprite memory directly to the panel
panel.pushColors(0, 0, WIDTH, HEIGHT, (uint16_t *)fbuff.getPointer());
```

If you don't use the TFT_eSPI then you're limited to much simpler primatives or using another library which I wasn't familiar with.

**Comment by bajacondor** [comment]
Hi James,
Thanks for your response!  I actually got this working just as you say.
I'm using the esprssif esp_display_panel lib so I'm pushing to the screen
with the drawBitmap() method.  I'm having a couple issues that I'll mention
in case you know of any better solutions:
1. The endian-ness of colors is reversed.  I'm having to shift all my
colors when rendering the final sprite.
2. Gradients are banded and look wrong.  I think this probably has to do
with #1 above.
3. Performance is pretty bad with larger sprites.  I was surprised that the
bottleneck is not drawing to the screen (it's a 480x480 RGB565) but the
layering of the sprites in memory.  I'm trying to do some fullscreen
rotation, so the main sprite has to be a bit bigger than the screensize,
then I use pushRotated() to another screen-size sprite, which then
renders.  I'm only getting ~4 fps with this method.  What kind of fps do
you see?  Is there anything I can do to speed things up?

Thank you!!
